src/query/functions/tests/it/type_check.rs

```diff
@@ -0,0 +1,147 @@
+// Copyright 2021 Datafuse Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use databend_common_expression::types::*;
+use databend_common_expression::FromData;
+use goldenfile::Mint;
+
+use crate::scalars::run_ast;
+
+#[test]
+fn test_type_check() {
+    let mut mint = Mint::new("tests/it/type_check/testdata");
+
+    let columns = [
+        ("s", StringType::from_data(vec!["s"])),
+        ("n_s", StringType::from_data(vec!["s"]).wrap_nullable(None)),
+        ("i8", Int8Type::from_data(vec![0])),
+        ("n_i8", Int8Type::from_data(vec![0]).wrap_nullable(None)),
+        ("u8", UInt8Type::from_data(vec![0])),
+        ("n_u8", UInt8Type::from_data(vec![0]).wrap_nullable(None)),
+        ("i16", Int16Type::from_data(vec![0])),
+        ("n_i16", Int16Type::from_data(vec![0]).wrap_nullable(None)),
+        ("u16", UInt16Type::from_data(vec![0])),
+        ("n_u16", UInt16Type::from_data(vec![0]).wrap_nullable(None)),
+        ("i32", Int32Type::from_data(vec![0])),
+        ("n_i32", Int32Type::from_data(vec![0]).wrap_nullable(None)),
+        ("u32", UInt32Type::from_data(vec![0])),
+        ("n_u32", UInt32Type::from_data(vec![0]).wrap_nullable(None)),
+        ("i64", Int64Type::from_data(vec![0])),
+        ("n_i64", Int64Type::from_data(vec![0]).wrap_nullable(None)),
+        ("u64", UInt64Type::from_data(vec![0])),
+        ("n_u64", UInt64Type::from_data(vec![0]).wrap_nullable(None)),
+        ("f32", Float32Type::from_data(vec![0.0])),
+        (
+            "n_f32",
+            Float32Type::from_data(vec![0.0]).wrap_nullable(None),
+        ),
+        ("f64", Float64Type::from_data(vec![0.0])),
+        (
+            "n_f64",
+            Float64Type::from_data(vec![0.0]).wrap_nullable(None),
+        ),
+        ("b", BooleanType::from_data(vec![true])),
+        (
+            "n_b",
+            BooleanType::from_data(vec![true]).wrap_nullable(None),
+        ),
+        ("d", DateType::from_data(vec![0])),
+        ("n_d", DateType::from_data(vec![0]).wrap_nullable(None)),
+        ("ts", TimestampType::from_data(vec![0])),
+        (
+            "n_ts",
+            TimestampType::from_data(vec![0]).wrap_nullable(None),
+        ),
+        // BinaryColumn != BinaryColumn
```

Comment:
thread 'type_check::test_type_check' panicked at src/query/functions/tests/it/scalars/mod.rs:227:5:
assertion `left == right` failed: arrow conversion went wrong
  left: BinaryColumn { data: 0x7b7d, offsets: [0, 2] }
 right: BinaryColumn { data: 0x7b7d, offsets: [0, 2] }